### PR TITLE
Dev/statement performance analyzer

### PR DESCRIPTION
**Pull request by JesperWisborgKrogh**
Create a report of the statements running on the server.
The views are calculated based on the overall and/or delta activity.

### CODE REVIEW

File: procedures/statement_performance_analyzer.sql [hunk 1]
```diff
@@ -0,0 +1,792 @@
...
+    -- Temporary table are used - disable sql_log_bin if necessary to prevent them replicating
+    SET @log_bin := @@sql_log_bin;
+    IF (@log_bin = 1) THEN
+        SET sql_log_bin = 0;
+    END IF;
+
+
+    -- Set configuration options
+    IF (@sys.statement_analyzer_limit IS NULL) THEN
+        SET @sys.statement_analyzer_limit = sys.sys_get_config('statement_analyzer_limit', '100');
```
Comment (MarkLeith) [inline procedures/statement_performance_analyzer.sql]: This should be consistent like the diagnostics routine variables - i.e it should prefix with the routine name like:

statement_performance_analyzer.limit

File: procedures/statement_performance_analyzer.sql [hunk 2]
```diff
@@ -0,0 +1,792 @@
...
+            PREPARE stmt_select_view FROM @sys.tmp.SQL;
+            EXECUTE stmt_select_view;
+            DEALLOCATE PREPARE stmt_select_view;
+        END IF;
+
+        IF (FIND_IN_SET('custom', in_views)) THEN
+            SELECT CONCAT('Top ', @sys.statement_analyzer_limit, ' Queries Using Custom View') AS 'Next Output';
+
+            IF (@sys.statement_analyzer_view IS NULL) THEN
+                SET @sys.statement_analyzer_view = sys.sys_get_config('sys.statement_analyzer_view', NULL);
```
Comment (MarkLeith) [inline procedures/statement_performance_analyzer.sql]: Same here with consistency in naming. And lack of sys_config row for both of the variables too.

File: README.md [hunk 3]
```diff
@@ -4399,6 +4399,207 @@ mysql> CALL sys.ps_truncate_all_tables(false);
...
++-------------------------------------+
+1 row in set (0.09 sec)
+
+...
+
+
+-- Use a custom view showing the top 10 query sorted by total execution time refreshing the view every minute using
+-- the watch command in Linux.
+
+mysql> mysql> CREATE OR REPLACE VIEW mydb.my_statements AS
```
Comment (MarkLeith) [inline README.md]: Odd couple of prompts here

File: procedures/statement_performance_analyzer.sql [hunk 4]
```diff
@@ -0,0 +1,792 @@
...
+        DEALLOCATE PREPARE stmt_create_table;
+
+    ELSEIF (in_action = 'save') THEN
+        SET @sys.tmp.SQL = CONCAT('DELETE FROM ', v_quoted_table);
+        IF (@sys.debug = 'ON') THEN
+            SELECT @sys.tmp.SQL AS 'Debug';
+        END IF;
+        PREPARE stmt_delete FROM @sys.tmp.SQL;
+        EXECUTE stmt_delete;
+        DEALLOCATE PREPARE stmt_delete;
```
Comment (MarkLeith) [inline procedures/statement_performance_analyzer.sql]: This pattern is repeated over and over, it seems like this is one of those cases where we could extract this to something like an "execute_prepared_statement" procedure, and just pass in the SQL string to prepare. I would do something like:

CALL execute_prepared_stmt(CONCAT('DELETE FROM ', v_quoted_table));

And have the routine itself set the input string in to a user variable, check debug, prepare/execute/deallocate.

File: procedures/statement_performance_analyzer.sql [hunk 5]
```diff
@@ -0,0 +1,792 @@
...
+            IF (in_table IS NULL) THEN
+                SET v_digests_table = 'tmp_digests';
+            ELSE
+                SET v_digests_table = v_quoted_table;
+            END IF;
+        ELSE
+            SET v_digests_table = 'tmp_digests_delta';
+            DROP TEMPORARY TABLE IF EXISTS tmp_digests_delta;
+            CREATE TEMPORARY TABLE tmp_digests_delta LIKE tmp_digests;
+            SET @sys.tmp.SQL = CONCAT('INSERT INTO tmp_digests_delta
```
Comment (MarkLeith) [inline procedures/statement_performance_analyzer.sql]: For larger statements like this, the execute_prepared_stmt procedure could still take this user variable (just for readability really), and still just allocate that to another internal user variable.

It also shows that the size of the query could be pretty large too, so make sure that the execute_prepared_stmt procedure *also* nulls out the temp user variable after it's been used.

File: procedures/statement_performance_analyzer.sql [hunk 6]
```diff
@@ -0,0 +1,792 @@
...
+                    -- No . in the table name - use current database
+                    -- DATABASE() will be the database of the procedure
+                    SET v_custom_db   = DATABASE(),
+                        v_custom_name = @sys.statement_analyzer_view;
+                ELSE
+                    SET v_custom_db   = SUBSTRING_INDEX(@sys.statement_analyzer_view, '.', 1);
+                    SET v_custom_name = SUBSTRING(@sys.statement_analyzer_view, CHAR_LENGTH(v_custom_db)+2);
+                END IF;
+
+                CALL table_exists(v_custom_db, v_custom_name, v_custom_view_exists);
```
Comment (MarkLeith) [inline procedures/statement_performance_analyzer.sql]: Prefix this with "sys.table_exists", these can be called from outside of the sys schema.

File: procedures/statement_performance_analyzer.sql [hunk 7]
```diff
@@ -0,0 +1,792 @@
...
+
+
+    IF (in_views IS NULL OR in_views = '') THEN
+        -- Set to default
+        SET in_views = 'with_runtimes_in_95th_percentile,analysis,with_errors_or_warnings,with_full_table_scans,with_sorting,with_temp_tables';
+    END IF;
+
+
+    -- Validate settings
+    CALL table_exists(DATABASE(), 'tmp_digests', v_tmp_digests_table_exists);
```
Comment (MarkLeith) [inline procedures/statement_performance_analyzer.sql]: Prefix this with "sys.table_exists", these can be called from outside of the sys schema.

File: procedures/statement_performance_analyzer.sql [hunk 8]
```diff
@@ -0,0 +1,792 @@
...
+            SELECT CONCAT('in_table is: db = ''', v_table_db, ''', table = ''', v_table_name, '''') AS 'Debug';
+        END IF;
+
+        IF (v_table_db = DATABASE() AND (v_table_name = 'tmp_digests' OR v_table_name = 'tmp_digests_delta')) THEN
+            SET v_error_msg = CONCAT('Invalid value for in_table: ', v_quoted_table, ' is reserved table name.');
+            SIGNAL SQLSTATE '45000'
+               SET MESSAGE_TEXT = v_error_msg;
+        END IF;
+
+        CALL table_exists(v_table_db, v_table_name, v_table_exists);
```
Comment (MarkLeith) [inline procedures/statement_performance_analyzer.sql]: Prefix this with "sys.table_exists", these can be called from outside of the sys schema.

**Comment by JesperWisborgKrogh** [comment]
Have made the following changes as requested:

* Renamed @sys.statement_analyzer_limit to @sys.statement_performance_analyzer.limit
* Renamed @sys.statement_analyzer_view to @sys.statement_performance_analyzer.view
* Updated README.md
* Added execute_prepared_stmt() to replace the repeated manual steps
* Prefixed table_exists() with "sys."

**Pull request merged**